webapp/store/selectors/nav.js

```diff
@@ -0,0 +1,176 @@
+
+      // get plugin path, if none, use name
+      let pathName = plugin.pathName ? plugin.pathName : plugin.name;
+
+      // sanitize of any other uri constructions like other nested paths
+      pathName = pathName.slice(pathName.lastIndexOf('/') + 1);
+
+      // make nested path
+      pathName = `${parentPath}/${pathName}`;
+      plugin.pathName = pathName;
+    }
+    updated.push(plugin);
+    return updated;
+  }, []);
+}
+
+/* Get plugin metadata list w/ unique pathNames
+ * @param {Object{}} metadata - object of plugin metadata objects
+ * @returns {Array{}} plugins - returns an array of metadata objects with unique paths
+ */
+export const getUniquePaths = metadata => {
+  assert(Array.isArray(metadata), 'Must pass array of metadata');
+  const paths = new Set();
+  const names = new Set();
+
+  return metadata.reduce((acc, plugin) => {
+    const { pathName, name, displayName: _displayName } = plugin;
+    if (pathName) {
+      let path = pathName;
+
+      if (paths.has(path)) {
+        // find unique suffix by incrementing counter
+        let counter = 1;
+        while (paths.has(`${path}-${counter}`)) {
+          counter++;
+        }
+        path = `${path}-${counter}`;
+      }
+
+      paths.add(path);
+      // set the metadata to the unique path
+      plugin.pathName = path;
+    }
+
+    // get unique displayName using same mechanism
+    let displayName = _displayName ? _displayName : name;
+    if (names.has(displayName)) {
+      let counter = 1;
+      while (names.has(`${displayName}-${counter}`)) {
+        counter++;
+      }
+      displayName = `${displayName}-${counter}`;
+    }
+    names.add(displayName);
+    plugin.displayName = displayName;
+
+    acc.push(plugin);
+    return acc;
+  }, []);
+};
+
+export function getNavItems(metadata = []) {
+  assert(Array.isArray(metadata), 'Must pass array to getNavItems');
+  return metadata.filter(plugin => plugin.nav || plugin.sidebar);
+}
+
+// Properly compose metadata for navigation
+// sort list of metadata -> compose nested paths -> make paths unique
+function composeNavItems(_navItems = []) {
+  assert(Array.isArray(_navItems), 'Must pass array to composeNavItems');
+  let navItems = getNavItems(_navItems);
+  navItems = sortPluginMetadata(navItems);
+  navItems = getNestedPaths(navItems);
+  navItems = getUniquePaths(navItems);
+  return navItems;
+}
+
+// selectors for converting pluginMetadata to nav list
+export const sortedNavItems = createSelector(getMetadataList, composeNavItems);
+
+// selector for ordering and composing sidebar navigation items
+export const sortedSidebarItems = createSelector(
+  getSidebarItems,
+  composeNavItems
+);
+
+// Useful selector for the Panel which just needs to match names to paths
+export const uniquePathsByName = createSelector(getMetadataList, metadata => {
+  const paths = {};
+  metadata
+    .filter(plugin => plugin.pathName)
+    .forEach(({ name, pathName }) => (paths[name] = pathName));
+  return paths;
```

Comment:
More of a nit as this works, but I'd lean towards chaining a filter into a reduce here personally

Reply:
good call. Cleaner that way.